### PR TITLE
[MRG] Fix pickle support for ArrayMemmapReducer

### LINKED ISSUE #629: Regression: TypeError: can't pickle weakref objects
In an attempt to solve some impossible to debug deadlocks I was getting with the `sklearn` version of `joblib`, I decided to test with the new dev version and the `loky` backend, however, when running my code it gives me an error:
`TypeError: can't pickle weakref objects`
The error happens with both the `loky` and `multiprocessing` backends (joblib master branch pulled from git earlier today), the same function was "working" (apart from the deadlocks, but that is a separate issue) in the previous version of `joblib`.

Are there any changes that make the code backwards incompatible? I checked the docs, but I didn't find anything useful there. Shouldn't the `cloudpickle` backend work fine with everything that was working before?

### PR DESCRIPTION
**Pull request by ogrisel**
Fix #629.

### CODE REVIEW

File: joblib/_memmapping_reducer.py [hunk 1]
```diff
@@ -282,6 +285,19 @@ def __init__(self, max_nbytes, temp_folder, mmap_mode, verbose=0,
         self._prewarm = prewarm
         self._memmaped_arrays = _WeakArrayKeyMap()
 
+    def __reduce__(self):
+        # The ArrayMemmapReducer is passed to the children processes: it need
```
Comment (aabadie) [inline joblib/_memmapping_reducer.py]: need**s** and also on the line below

File: joblib/test/test_parallel.py [hunk 2]
```diff
@@ -875,6 +875,13 @@ def generate_arrays(n):
     for result, expected in zip(results, generate_arrays(len(results))):
         np.testing.assert_array_equal(expected, result)
 
+    # Second call to force loky to adapt the executor by growing the number
+    # of worker processes. This is a non regression test for: #629.
```
Comment (aabadie) [inline joblib/test/test_parallel.py]: Maybe put the full url of the issue ?

File: joblib/_memmapping_reducer.py [hunk 3]
```diff
@@ -282,6 +285,19 @@ def __init__(self, max_nbytes, temp_folder, mmap_mode, verbose=0,
...
+        # to be pickled but the _WeakArrayKeyMap need to be skipped as it's
+        # only guaranteed to be consistent with the parent process memory
+        # garbage collection.
+        args = (self._max_nbytes, self._temp_folder,
```
Comment (aabadie) [inline joblib/_memmapping_reducer.py]: I think all args could be on the same line

File: joblib/test/test_memmapping.py [hunk 4]
```diff
@@ -3,6 +3,8 @@
 import sys
 import platform
 import gc
+import pickle
+import pytest
```
Comment (aabadie) [inline joblib/test/test_memmapping.py]: I would use from `joblib.testing import raises` which a wrapper around `pytest.raises`

**Comment by ogrisel** [comment]
Ok the sphinx issue is unrelated and should be fixed as part of a separate PR. @aabadie thanks for the review!

**Pull request merged**